### PR TITLE
S3 basic implementation with AWS SDK 1.x

### PR DESCRIPTION
**Pull request by gspinaci**

### CODE REVIEW

File: src/main/java/org/researchspace/services/storage/s3/S3Storage.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        } catch (Exception e) {
+
+            throw new StorageException(e.getMessage());
+        }
+    }
+
+    @Override
+    public ObjectRecord appendObject(StoragePath path, ObjectMetadata metadata, InputStream content, long contentLength)
+            throws StorageException {
+        // TODO Auto-generated method stub
```
Comment (aindlq) [inline src/main/java/org/researchspace/services/storage/s3/S3Storage.java:181]: can we have implementation for append and delete, I believe it should be quite straightforward to do:

https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3.html#putObject-java.lang.String-java.lang.String-java.io.InputStream-com.amazonaws.services.s3.model.ObjectMetadata-

https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3.html#deleteObject-java.lang.String-java.lang.String-

File: src/main/java/org/researchspace/services/storage/file/NonVersionedFileStorage.java [hunk 2]
```diff
@@ -40,31 +41,62 @@
...
     public NonVersionedFileStorage(PathMapping paths, Config config) {
-        this.paths = paths;
+
         this.config = config;
+
+        // Check if the config requires a hierarchical path mapping
+        String pathMapping = config.getPathMapping();
+        IRI pathMappingBaseIri = config.getPathMappingBaseIri();
+
+        if (Objects.nonNull(pathMapping) && Objects.nonNull(pathMappingBaseIri) && pathMapping.equals("hierarchical")) {
```
Comment (aindlq) [inline src/main/java/org/researchspace/services/storage/file/NonVersionedFileStorage.java:68]: I believe it is a wrong place for that, probably a better one is https://github.com/researchspace/researchspace/blob/3b072f67f245e77421a87842d1c02afe5652c45d/src/main/java/org/researchspace/services/storage/MainPlatformStorage.java#L180 . So if you move this logic there and also move pathMapping and pathMappingBaseIri into StorageConfig class we can reuse the same logic in the S3Storage and in the GitStorage without any additional effort

**Review by aindlq — CHANGES_REQUESTED** [review]

File: src/main/java/org/researchspace/services/storage/file/NonVersionedFileStorage.java (hunk 2 above)
Comment (gspinaci) [reply src/main/java/org/researchspace/services/storage/file/NonVersionedFileStorage.java]: The last commit has been moved to a new proper branch: https://github.com/researchspace/researchspace/tree/hierarchical.

**Comment by gspinaci** [comment]
I'm also creating unit tests

File: src/main/java/org/researchspace/services/storage/s3/S3Storage.java [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+            String key = getKeyFromStoragePath(path);
+
+            S3StorageLocation location = new S3StorageLocation(config.getBucket(), key);
+
+            // TODO: calculate revision
+            objectRecord = new ObjectRecord(location, path, "", metadata);
+
+            // TODO: check PutObjectResult
+            s3.putObject(config.getBucket(), key, content.toString());
```
Comment (aindlq) [inline src/main/java/org/researchspace/services/storage/s3/S3Storage.java:231]: It doesn't work, you probably need to use other method that takes input stream and metadata - https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3Client.html#putObject-java.lang.String-java.lang.String-java.io.InputStream-com.amazonaws.services.s3.model.ObjectMetadata-

File: src/main/resources/org/researchspace/apps/help/data/templates/http%3A%2F%2Fhelp.researchspace.org%2Fresource%2FStorage.html [hunk 4]
```diff
@@ -163,8 +164,8 @@ <h3 id="runtime-s3-example">Example: S3 as <code>runtime</code> Storage</h3>
 -Dconfig.storage.runtime.endpoint=https://s3.amazonaws.com/ \
 -Dconfig.storage.runtime.signingRegion=us-east-1 \
```
Comment (aindlq) [inline src/main/resources/org/researchspace/apps/help/data/templates/http%3A%2F%2Fhelp.researchspace.org%2Fresource%2FStorage.html:165]: I think it is just `region` not `signingRegion`, could you please change it in the docs

File: src/main/java/org/researchspace/services/storage/s3/S3Storage.java (hunk 3 above)
Comment (gspinaci) [reply src/main/java/org/researchspace/services/storage/s3/S3Storage.java]: Fix in #5fa15d7

File: src/main/resources/org/researchspace/apps/help/data/templates/http%3A%2F%2Fhelp.researchspace.org%2Fresource%2FStorage.html (hunk 4 above)
Comment (gspinaci) [reply src/main/resources/org/researchspace/apps/help/data/templates/http%3A%2F%2Fhelp.researchspace.org%2Fresource%2FStorage.html]: Fix in https://github.com/researchspace/researchspace/pull/207/commits/5fa15d72fb75923c535899b1c3da3056efa03f90

File: src/main/java/org/researchspace/services/storage/s3/S3Storage.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                Optional<ObjectRecord> record = getObject(path.get(), null);
+
+                if (record.isPresent())
+                    records.add(record.get());
+            }
+
+            return records;
+    
+        } catch (Exception e) {
+
```
[the remark this answers is not in the document]
Comment (gspinaci) [reply src/main/java/org/researchspace/services/storage/s3/S3Storage.java:191]: Fix in https://github.com/researchspace/researchspace/pull/207/commits/b0de8953e9ed425654b3eca2b1823c88dd10a59f

**Comment by gspinaci** [comment]
The PR has been moved to #256

**Pull request closed without merging**